KappaAnalysis/src/Producers/GenDiLeptonDecayModeProducer.cc

```diff
@@ -38,68 +38,30 @@ void GenDiLeptonDecayModeProducer::Init(KappaSettings const& settings)
                                            KappaSettings const& settings) const
 {
-	assert(event.m_genParticles);
-
-	for (KGenParticles::const_iterator genParticle = event.m_genParticles->begin();
-		 genParticle != event.m_genParticles->end(); ++genParticle)
+	std::map<int, int> nDecayProductsPerType;
+	size_t iDaughter = 0;
+	for (std::vector<KGenParticle*>::iterator genParticle = product.m_genLeptonsFromBosonDecay.begin();
+	     genParticle != product.m_genLeptonsFromBosonDecay.end() && (iDaughter < 2); ++genParticle)
 	{
-		if (Utility::Contains(settings.GetBosonPdgIds(), std::abs(genParticle->pdgId)) && Utility::Contains(settings.GetBosonStatuses(), genParticle->status()))
-		{
-			std::map<int, int> nDecayProductsPerType;
-			product.m_genDiLeptonBoson = (*genParticle).p4;
-
-			for (std::vector<unsigned int>::const_iterator decayParticleIndex = genParticle->daughterIndices.begin();
-			     decayParticleIndex != genParticle->daughterIndices.end(); ++decayParticleIndex)
-			{
-				int pdgId = std::abs(event.m_genParticles->at(*decayParticleIndex).pdgId);
-				nDecayProductsPerType[pdgId] = SafeMap::GetWithDefault(nDecayProductsPerType, pdgId, 0) + 1;
-			}
-
-			if (SafeMap::GetWithDefault(nDecayProductsPerType, DefaultValues::pdgIdElectron, 0) >= 2)
-			{
-				product.m_genDiLeptonDecayMode = KappaEnumTypes::DiLeptonDecayMode::EE;
-			}
-			else if (SafeMap::GetWithDefault(nDecayProductsPerType, DefaultValues::pdgIdMuon, 0) >= 2)
-			{
-				product.m_genDiLeptonDecayMode = KappaEnumTypes::DiLeptonDecayMode::MM;
-			}
-			else if (SafeMap::GetWithDefault(nDecayProductsPerType, DefaultValues::pdgIdTau, 0) >= 2)
-			{
-				product.m_genDiLeptonDecayMode = KappaEnumTypes::DiLeptonDecayMode::TT;
-			}
-			else
-			{
-				product.m_genDiLeptonDecayMode = KappaEnumTypes::DiLeptonDecayMode::NONE;
-			}
-
-			// exit loop after one boson is found
-			if (product.m_genDiLeptonDecayMode != KappaEnumTypes::DiLeptonDecayMode::NONE)
-			{
-				break;
-			}
-		}
+		int pdgId = std::abs((*genParticle)->pdgId);
+		nDecayProductsPerType[pdgId] = SafeMap::GetWithDefault(nDecayProductsPerType, pdgId, 0) + 1;
+		++iDaughter;
 	}
 	
-	// If no boson has been found in the event, try to reconstruct it from the first two decay
-	// products available in the list of gen. particles
-	if (product.m_genDiLeptonDecayMode == KappaEnumTypes::DiLeptonDecayMode::NONE)
+	if (SafeMap::GetWithDefault(nDecayProductsPerType, DefaultValues::pdgIdElectron, 0) >= 2)
+	{
+		product.m_genDiLeptonDecayMode = KappaEnumTypes::DiLeptonDecayMode::EE;
+	}
+	else if (SafeMap::GetWithDefault(nDecayProductsPerType, DefaultValues::pdgIdMuon, 0) >= 2)
+	{
+		product.m_genDiLeptonDecayMode = KappaEnumTypes::DiLeptonDecayMode::MM;
+	}
+	else if (SafeMap::GetWithDefault(nDecayProductsPerType, DefaultValues::pdgIdTau, 0) >= 2)
+	{
+		product.m_genDiLeptonDecayMode = KappaEnumTypes::DiLeptonDecayMode::TT;
+	}
+	else
 	{
-		size_t iDaughter = 0;
```

Comment:
ok found it
https://github.com/artus-analysis/Artus/blob/master/KappaAnalysis/src/Producers/GenBosonProducers.cc#L78-L101
I was wondering: For which processes/ production campaigns is this actually needed? And how one be sure that the the first two leptons one finds are the Boson decay products?
cheers